### ISSUE TITLE
Memory leak in Conv1d

### ISSUE THREAD

**Issue by hoagy-davis-digges** [pytorch/pytorch#98688]
### 🐛 Describe the bug

There appears to be a memory leak in conv1d, when I run the following code the cpu ram usage ticks up continually, if I remove  x = self.conv1(x) this no longer happens

```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import IterableDataset, DataLoader
import numpy as np

# 1. Import the necessary libraries

# 2. Define the custom dataset class
class CustomDataset(IterableDataset):
    def __iter__(self):
        return self

    def __next__(self):
        rows = np.random.randint(100, 300)
        columns = np.random.randint(200, 400)
        data = np.random.rand(rows, columns, 100)
        labels = np.random.randint(0, 17, [rows, columns])
        return torch.tensor(data, dtype=torch.float32), torch.tensor(labels)


# 3. Define the model architecture
class Conv1DModel(nn.Module):
    def __init__(self, in_channels, out_channels, kernel_size, padding):
        super(Conv1DModel, self).__init__()
        self.conv1 = nn.Conv1d(in_channels,
                               out_channels,
                               kernel_size=kernel_size,
                               padding=padding)
        self.relu = nn.ReLU()
        self.fc = nn.Linear(out_channels, 17)

    def forward(self, x):
        x = x.transpose(1, 2)
        x = self.conv1(x) # commenting this removes leak
        x = x.transpose(1, 2)
        x = self.relu(x)
        x = self.fc(x)
        return x


# Create the dataset and data loader
dataset = CustomDataset()

# Initialize the model
model = Conv1DModel(in_channels=100, out_channels=100, kernel_size=1, padding=0)

# Set loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Train the model
num_epochs = 10
model.to('cuda')

for epoch in range(num_epochs):
    for i, (inputs, targets) in enumerate(dataset):
        outputs = model(inputs.to('cuda'))
        loss = criterion(outputs.transpose(1, 2), targets.to('cuda'))
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        print(f"Batch {i} Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}")
```

### Versions

Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.26.1
Libc version: glibc-2.37

Python version: 3.9.10 (main, Dec 20 2022, 13:59:33)  [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-6.1.22-1-lts-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 12.1.66
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Super with Max-Q Design
Nvidia driver version: 530.41.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) i7-10875H CPU @ 2.30GHz
CPU family:                      6
Model:                           165
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        2
Frequency boost:                 enabled
CPU(s) scaling MHz:              61%
CPU max MHz:                     2301.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4601.60
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        2 MiB (8 instances)
L3 cache:                        16 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Vulnerable: No microcode
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==4.0.1
[pip3] numpy==1.23.1
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-triton==2.1.0+46672772b4
[pip3] torch==2.0.0
[pip3] torchaudio==2.1.0.dev20230407+cu118
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0

cc @ezyang @gchanan @zou3519

**Comment by zou3519** [comment]
Tentatively marking hi-pri for memory leak, but we need to reproduce this on our side

**Comment by malfet** [comment]
Leaving as high-pri to repo

**Comment by ptrblck** [comment]
No reproduction in `2.0.0+cu117` and `2.0.0+cu118` on a 3090 as both show a constant memory usage via `nvidia-smi`.

**Comment by zou3519** [comment]
@ptrblck was the CPU memory usage also constant?

**Comment by hoagy-davis-digges** [comment]
To clarify it was CPU memory usage that was increasing

**Comment by mingfeima** [comment]
@chunyuan-w , could you please take a look at this issue ?

**Comment by chunyuan-w** [comment]
> our validation team has verified that this issue is reproducible on CPU, will have it fixed!
> 
> @chunyuan-w , could you please take a look at this issue ?

Sure, let me check.

**Comment by chunyuan-w** [comment]
When our validation team tried to reproduce it on CPU, all the `to('cuda')` has been changed to `to('cpu')`.
As it is a case with dynamic shape inputs, the primitive and primitive description cache has made the memory increase at the beginning when it hasn't reached the maximum cache capacity size. By setting the below environment variables to decrease the cache capacity, the memory won't increase anymore:
`ONEDNN_PRIMITIVE_CACHE_CAPACITY=0  LRU_CACHE_CAPACITY=1`

Using the original script with `to('cuda)`, our validation team fails to reproduce the issue using `2.1.0.dev20230330` or the below version. The CPU memory usage won't increase continually.
```
pytorch                   2.0.0           py3.10_cuda11.7_cudnn8.5.0_0    pytorch
pytorch-cuda              11.7                 h778d358_3    pytorch
```

**Comment by gaelm** [comment]
hello, same issue for us.
it seems related to the other issue https://github.com/pytorch/pytorch/issues/101921
we have copied their example:
```
import random
import resource
import torch
from torch import nn
import os

os.environ["LRU_CACHE_CAPACITY"] = "1"
os.environ["ONEDNN_PRIMITIVE_CACHE_CAPACITY"] = "0"


class ToyNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(in_channels=16, out_channels=64, kernel_size=11, padding=5)
        self.conv1 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=11, padding=5)
        self.conv2 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=11, padding=5)
        self.conv3 = nn.Conv2d(in_channels=64, out_channels=64, kernel_size=11, padding=5)

    def forward(self, x):
        x = self.conv(x)
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.conv3(x)
        return x


device = torch.device("cuda:0")
model = ToyNet().to(device)

max_dim0, max_dim1 = 0, 0
for i in range(500):
    max_mem = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024  # MB
    max_cuda_mem = torch.cuda.max_memory_allocated(device) / 1024 / 1024  # MB

    dim = (random.randint(128, 512), random.randint(128, 512))
    model(torch.randn((32, 16, *dim), device=device))

    max_dim0 = max(max_dim0, dim[0])
    max_dim1 = max(max_dim1, dim[1])
    print(f"{i}: max dim: ({max_dim0}, {max_dim1}), CPU: {max_mem:.3f}MB, GPU: {max_cuda_mem:.3f}MB")
```
the environment variables don't help.
An example of final output is:
```
461: max dim: (512, 511), CPU: 2109.723MB, GPU: 8625.545MB
462: max dim: (512, 511), CPU: 2109.723MB, GPU: 8625.545MB
463: max dim: (512, 511), CPU: 2109.980MB, GPU: 8625.545MB
464: max dim: (512, 511), CPU: 2109.980MB, GPU: 8625.545MB
465: max dim: (512, 511), CPU: 2109.980MB, GPU: 8625.545MB
466: max dim: (512, 511), CPU: 2109.980MB, GPU: 8625.545MB
467: max dim: (512, 511), CPU: 2109.980MB, GPU: 8625.545MB
468: max dim: (512, 511), CPU: 2110.238MB, GPU: 8625.545MB
469: max dim: (512, 511), CPU: 2110.238MB, GPU: 8625.545MB
470: max dim: (512, 511), CPU: 2110.238MB, GPU: 8625.545MB
471: max dim: (512, 511), CPU: 2110.238MB, GPU: 8625.545MB
472: max dim: (512, 511), CPU: 2110.238MB, GPU: 8625.545MB
473: max dim: (512, 511), CPU: 2110.496MB, GPU: 8625.545MB
474: max dim: (512, 511), CPU: 2110.496MB, GPU: 8625.545MB
475: max dim: (512, 511), CPU: 2110.496MB, GPU: 8625.545MB
476: max dim: (512, 511), CPU: 2110.496MB, GPU: 8625.545MB
477: max dim: (512, 511), CPU: 2110.754MB, GPU: 8625.545MB
478: max dim: (512, 511), CPU: 2110.754MB, GPU: 8625.545MB
479: max dim: (512, 511), CPU: 2110.754MB, GPU: 8625.545MB
480: max dim: (512, 511), CPU: 2110.754MB, GPU: 8625.545MB
481: max dim: (512, 511), CPU: 2110.754MB, GPU: 8625.545MB
482: max dim: (512, 511), CPU: 2111.012MB, GPU: 8625.545MB
483: max dim: (512, 511), CPU: 2111.012MB, GPU: 8625.545MB
484: max dim: (512, 511), CPU: 2111.012MB, GPU: 8625.545MB
485: max dim: (512, 511), CPU: 2111.012MB, GPU: 8625.545MB
486: max dim: (512, 511), CPU: 2111.270MB, GPU: 8625.545MB
487: max dim: (512, 511), CPU: 2111.270MB, GPU: 8625.545MB
488: max dim: (512, 511), CPU: 2111.270MB, GPU: 8625.545MB
489: max dim: (512, 511), CPU: 2111.270MB, GPU: 8625.545MB
490: max dim: (512, 511), CPU: 2111.527MB, GPU: 8625.545MB
491: max dim: (512, 511), CPU: 2111.527MB, GPU: 8625.545MB
492: max dim: (512, 511), CPU: 2111.527MB, GPU: 8625.545MB
493: max dim: (512, 511), CPU: 2111.527MB, GPU: 8625.545MB
494: max dim: (512, 511), CPU: 2111.527MB, GPU: 8625.545MB
495: max dim: (512, 511), CPU: 2111.785MB, GPU: 8625.545MB
496: max dim: (512, 511), CPU: 2111.785MB, GPU: 8625.545MB
497: max dim: (512, 511), CPU: 2111.785MB, GPU: 8625.545MB
498: max dim: (512, 511), CPU: 2111.785MB, GPU: 8625.545MB
499: max dim: (512, 511), CPU: 2112.043MB, GPU: 8625.545MB
```
the max GPU memory usage is stable but the max CPU memory keeps growing.

**Comment by ZailiWang** [comment]
The memleak is only observed if model inference executed on GPU.
No memleak observed if executed on CPU.

**Comment by eqy** [comment]
I can reproduce this and this is likely a consequence of cuDNN V8 API execution plans taking up more host memory than algo perf data did in V7. Some additional memory usage is expected with each dynamic shape as these execution plans are cached in case the same workload is reached (as is the case in #101921).

You can use the `TORCH_CUDNN_V8_API_DISABLED=1` environment variable as a workaround for this issue in the meantime.

**Comment by colesbury** [comment]
@eqy is there a plan to handle the increased memory usage?

**Comment by eqy** [comment]
@colesbury Yes, the current plan is to implement a LRU cache immediately after #104122 is merged as I wanted to clean up some handling of cache keys.

**Comment by eqy** [comment]
Closing as #104122 has been merged.

**Comment by chaseaucoin** [comment]
@eqy this seems to still be an issue in the most recent publicly available version of torch

**Comment by eqy** [comment]
@chaseaucoin could you post a repro please? Growing host memory usage is expected up to `TORCH_CUDNN_V8_LRU_CACHE_LIMIT` which is by default 10000 shapes.

**Comment by chaseaucoin** [comment]
It happened to me when using the Xtts models from Coqui. I have the reproduction steps here:

https://github.com/coqui-ai/TTS/issues/3463

If there is any additional information I can provide please let me know and I'll be happy to help.

**Comment by eqy** [comment]
@chaseaucoin this issue relates to host, or CPU memory usage when using the CUDNN V8 API, rather than VRAM or device memory which seems to be what your issue mentions. I'm not really familiar with Windows use-cases, but by default PyTorch (and by extension, cuDNN) is "free" to use any available memory for temporary workspaces. As the V8 API can enable kernels not visible under the V7 API, you may be hitting kernel(s) that are using large amounts of workspace memory which may not play nice with Windows. You can reduce the amount of workspace memory that CUDNN is allowed to use via e.g., `CUDNN_CONV_WSCAP_DBG=512` to limit the maximum workspace size used by cuDNN to 512MiB.

If however you mean to say that the host memory usage is causing your issue, then this is expected behavior unless you can show that you are using more than the default number of shapes that cuDNN is allowed to cache: 10000. You can try setting this environment variable to a lower number (e.g., `TORCH_CUDNN_V8_LRU_CACHE_LIMIT=100` or `TORCH_CUDNN_V8_LRU_CACHE_LIMIT=1000` to see if that improves the memory usage behavior (though this may come at a hit to performance).

**Comment by chaseaucoin** [comment]
@eqy  The CPU memory isn't getting pegged. All available GPU VRAM gets consumed within 1 second until the system as a whole shuts down. That feels like a bug not a feature if it does that by default. TORCH_CUDNN_V8_API_DISABLED=1 seems to stop it so that feels like it is ultimately as issue with the V8 API.

**Comment by eqy** [comment]
@chaseaucoin In that case please consider opening a new issue for follow-ups as this issue is for host (CPU) memory usage under the V8 API.

PyTorch in this case wouldn't have visibility into what is "polite" other system processes (such as the display), and thus it's expected behavior for a _temporary_ workspace to be able to use all available GPU memory if necessary. As PyTorch uses a caching allocator, this "temporary" memory may be held on by the caching allocator unless explicitly cleared via `torch.cuda.empty_cache`. This greedy allocation behavior may not seem "nice" to other processes on the system, but it is acceptable in the common use-case of PyTorch being the only application running on a given GPU.

What you are observing is likely an artifact of additional kernels that use larger temporary workspaces being visible and selected under the V8 API. We may consider adding a flag for user convenience to limit the maximum workspace size that cuDNN is allowed to use, but at present this is expected behavior unless you can demonstrate that your problem still occurs when ``CUDNN_CONV_WSCAP_DBG` is set to a small value or when [`torch.cuda.set_per_process_memory_fraction`](https://pytorch.org/docs/stable/generated/torch.cuda.set_per_process_memory_fraction.html) is set to a small value.
